File: apps/api/src/modules/proxy/cache.ts
import { createHash } from "node:crypto";
import type { Database } from "@raven/db";
import type { Redis } from "ioredis";
import { analyzeContent } from "./content-analyzer";
import type { GuardrailMatch } from "./guardrails";
import { logAndPublish, updateLastUsed } from "./logger";
import { extractCachedUsage, extractModel } from "./usage-mapper";

const DEFAULT_TTL_SECONDS = 3600;

export interface CacheCheckResult {
  hit: true;
  body: string;
  parsed: Record<string, unknown>;
}

export interface CacheMissResult {
  hit: false;
}

export type CacheResult = CacheCheckResult | CacheMissResult;

const buildCacheKey = (
  orgId: string,
  provider: string,
  model: string,
  body: Record<string, unknown>
): string => {
  const content = body.messages ?? body.input ?? [];
  const temperature = body.temperature ?? null;
  const system = body.system ?? null;
  const tools = body.tools ?? null;
  const payload = `${orgId}:${provider}:${model}:${JSON.stringify(content)}:${temperature}:${JSON.stringify(system)}:${JSON.stringify(tools)}`;
  const hash = createHash("sha256").update(payload).digest("hex");
  return `cache:resp:${hash}`;
};

export const checkCache = async (
  redis: Redis,
  orgId: string,
  provider: string,
  requestBody: Record<string, unknown>
): Promise<CacheResult> => {
  if (requestBody.stream === true) {
    return { hit: false };
  }

  const model = (requestBody.model as string) ?? "unknown";
  const key = buildCacheKey(orgId, provider, model, requestBody);
  const cached = await redis.get(key);

  if (!cached) {
    return { hit: false };
  }

  let parsed: Record<string, unknown> = {};
  try {
    parsed = JSON.parse(cached);
  } catch {
    parsed = {};
  }

  return { body: cached, hit: true, parsed };
};

export const storeCache = async (
  redis: Redis,
  orgId: string,
  provider: string,
  requestBody: Record<string, unknown>,
  responseBody: string,
  ttlSeconds: number = DEFAULT_TTL_SECONDS
): Promise<void> => {
  if (requestBody.stream === true) {
    return;
  }

  const model = (requestBody.model as string) ?? "unknown";
  const key = buildCacheKey(orgId, provider, model, requestBody);
  await redis.set(key, responseBody, "EX", ttlSeconds);
};

/**
 * Build the response for a cache hit, logging the request and returning the cached body.
 * Shared by both the proxy handler and the OpenAI-compat handler.
 */
export const serveCacheHit = (
  db: Database,
  cacheResult: CacheCheckResult,
  opts: {
    guardrailMatches: GuardrailMatch[];
    guardrailWarnings: string[];
    method: string;
    model: string;
    organizationId: string;
    parsedBody: Record<string, unknown>;
    path: string;
    providerConfigId: string;
    providerName: string;
    redis: Redis;
    sessionHeader: string | null;
    startTime: number;
    teamId: string | null;
    virtualKeyId: string;
  }
): Response => {
  const latencyMs = Date.now() - opts.startTime;
  const analysis = analyzeContent(opts.parsedBody, opts.sessionHeader);
  const usage = extractCachedUsage(cacheResult.parsed);

  logAndPublish(
    db,
    {
      cachedTokens: usage.cachedTokens,
      cacheHit: true,
      cost: 0,
      guardrailMatches:
        opts.guardrailMatches.length > 0 ? opts.guardrailMatches : undefined,
      hasImages: analysis.hasImages,
      hasToolUse: analysis.hasToolUse,
      imageCount: analysis.imageCount,
      inputTokens: usage.inputTokens,
      latencyMs,
      method: opts.method,
      model: extractModel(cacheResult.parsed, opts.model),
      organizationId: opts.organizationId,
      outputTokens: usage.outputTokens,
      path: opts.path,
      provider: opts.providerName,
      providerConfigId: opts.providerConfigId,
      reasoningTokens: usage.reasoningTokens,
      requestBody: opts.parsedBody,
      sessionId: analysis.sessionId,
      statusCode: 200,
      toolCount: analysis.toolCount,
      toolNames: analysis.toolNames,
      virtualKeyId: opts.virtualKeyId
    },
    { redis: opts.redis, teamId: opts.teamId }
  );
  updateLastUsed(opts.redis, opts.virtualKeyId);

  const headers: Record<string, string> = {
    "content-type": "application/json"
  };
  if (opts.guardrailWarnings.length > 0) {
    headers["X-Guardrail-Warnings"] = opts.guardrailWarnings.join("; ");
  }
  return new Response(cacheResult.body, { headers, status: 200 });
};

File: apps/api-py/src/proxy/cache.py
import hashlib
import json
from typing import Any

from redis.asyncio import Redis
from sqlalchemy.ext.asyncio import AsyncSession

DEFAULT_TTL_SECONDS = 3600


def _build_cache_key(provider: str, model: str, body: dict[str, Any]) -> str:
    content = body.get("messages") or body.get("input") or []
    temperature = body.get("temperature")
    system = body.get("system")
    tools = body.get("tools")
    parts = [provider, model, json.dumps(content), str(temperature)]
    parts.extend([json.dumps(system), json.dumps(tools)])
    payload = ":".join(parts)
    hash_hex = hashlib.sha256(payload.encode("utf-8")).hexdigest()
    return f"cache:resp:{hash_hex}"


async def check_cache(
    redis: Redis, provider: str, request_body: dict[str, Any]
) -> dict:
    if request_body.get("stream") is True:
        return {"hit": False}

    model = request_body.get("model", "unknown")
    key = _build_cache_key(provider, model, request_body)
    cached = await redis.get(key)

    if not cached:
        return {"hit": False}

    try:
        parsed = json.loads(cached)
    except (json.JSONDecodeError, TypeError):
        parsed = {}

    return {"hit": True, "body": cached, "parsed": parsed}


async def store_cache(
    redis: Redis,
    provider: str,
    request_body: dict[str, Any],
    response_body: str,
    ttl_seconds: int = DEFAULT_TTL_SECONDS,
) -> None:
    if request_body.get("stream") is True:
        return

    model = request_body.get("model", "unknown")
    key = _build_cache_key(provider, model, request_body)
    await redis.set(key, response_body, ex=ttl_seconds)


def serve_cache_hit(
    *,
    session: AsyncSession,
    cache_result: dict,
    start_time: float,
    virtual_key_id: str,
    provider_name: str,
    provider_config_id: str,
    model: str,
    method: str,
    path: str,
    end_user: str | None,
    user_agent: str | None,
    session_id: str | None,
    redis: Redis,
    guardrail_warnings: list[str],
):
    from fastapi.responses import Response

    headers = {"Content-Type": "application/json"}
    if guardrail_warnings:
        headers["X-Guardrail-Warnings"] = "; ".join(guardrail_warnings)

    return Response(
        content=cache_result["body"],
        status_code=200,
        headers=headers,
    )
